### PR TITLE
except_regex argument for todo_comment_linter() to skip valid TODO comments

### LINKED ISSUE #2047: todo_comment_linter could get an argument for valid TODO patterns
This can allow `TODO` _only if_ they fit certain patterns.

This could be used to enforce a rule like "All TODO must be tied to a bug, not a person".

```r
## BAD
# TODO(michaelchirico): fix this
# TODO: add testing

## GOOD
# TODO(#nnn): fix this
# TODO(#mmm): add testing
```

Maybe an argument like `allow_regex` or `allow_style_regex`, in this case `allow_regex = "TODO[(][#][0-9]+[)]:"` could work.

### PR DESCRIPTION
**Pull request by MichaelChirico**
Based against #2438.

Closes #2047

### CODE REVIEW

**Review by AshesITR — DISMISSED** [review]
Enable it for us? (Follow up is fine)

**Comment by MichaelChirico** [comment]
Hmm that blew up. Let me know if anything here should be split into its own PR.

File: R/brace_linter.R [hunk 1]
```diff
@@ -74,7 +74,7 @@ brace_linter <- function(allow_single_line = FALSE) {
     )")
   ))
 
-  # TODO (AshesITR): if c_style_braces is TRUE, invert the preceding-sibling condition
+  # TODO(AshesITR): if c_style_braces is TRUE, invert the preceding-sibling condition
```
Comment (AshesITR) [inline R/brace_linter.R:77]:
```suggestion
  # TODO(#1103): if c_style_braces is TRUE, invert the preceding-sibling condition
```

File: R/brace_linter.R [hunk 2]
```diff
@@ -109,7 +109,7 @@ brace_linter <- function(allow_single_line = FALSE) {
     )"
   ))
 
-  # TODO (AshesITR): if c_style_braces is TRUE, skip the not(ELSE) condition
+  # TODO(AshesITR): if c_style_braces is TRUE, skip the not(ELSE) condition
```
Comment (AshesITR) [inline R/brace_linter.R:112]:
```suggestion
  # TODO(#1103): if c_style_braces is TRUE, skip the not(ELSE) condition
```

File: R/brace_linter.R [hunk 3]
```diff
@@ -121,7 +121,7 @@ brace_linter <- function(allow_single_line = FALSE) {
   xp_else_closed_curly <- "preceding-sibling::IF/following-sibling::expr[2]/OP-RIGHT-BRACE"
   # need to (?) repeat previous_curly_path since != will return true if there is
   #   no such node. ditto for approach with not(@line1 = ...).
-  # TODO (AshesITR): if c_style_braces is TRUE, this needs to be @line2 + 1
+  # TODO(AshesITR): if c_style_braces is TRUE, this needs to be @line2 + 1
```
Comment (AshesITR) [inline R/brace_linter.R:124]:
```suggestion
  # TODO(#1103): if c_style_braces is TRUE, this needs to be @line2 + 1
```

File: R/object_usage_linter.R [hunk 4]
```diff
@@ -87,8 +87,6 @@ object_usage_linter <- function(interpret_glue = TRUE, skip_with = TRUE) {
         skip_with = skip_with
       )
 
-      # TODO handle assignment functions properly
```
Comment (MichaelChirico) [inline R/object_usage_linter.R:90]: I don't know what this TODO is talking about, exactly. I tried these snippets and they lint as expected:

```r
lint(linters = object_usage_linter(), "
foo <- function() {
  not_exists(x) <- 2
  not_exists(x)
}")
# <text>:3:3: warning: [object_usage_linter] no visible global function definition for 'not_exists<-'
#   not_exists(x) <- 2
#   ^~~~~~~~~~
# <text>:4:3: warning: [object_usage_linter] no visible global function definition for 'not_exists'
#   not_exists(x)
#   ^~~~~~~~~~
lint(linters = object_usage_linter(), "
not_exists <- function(x) x
foo <- function() {
  not_exists(x) <- 2
  not_exists(x)
}")
# <text>:4:3: warning: [object_usage_linter] no visible global function definition for 'not_exists<-'
#   not_exists(x) <- 2
#   ^~~~~~~~~~
```

So deleting and waiting for something to be reported in tracker if there's actually something there.

File: R/object_usage_linter.R [hunk 5]
```diff
@@ -211,7 +209,7 @@ parse_check_usage <- function(expression,
   # nocov start
   is_missing <- is.na(res$message)
   if (any(is_missing)) {
-    # TODO (AshesITR): Remove this in the future, if no bugs arise from this safeguard
+    # TODO(AshesITR): Remove this in the future, if no bugs arise from this safeguard
```
Comment (AshesITR) [inline R/object_usage_linter.R:212]: Convert to issue?

File: R/todo_comment_linter.R [hunk 6]
```diff
@@ -3,22 +3,24 @@
...
-#'   linters = todo_comment_linter()
+#'   text = "pi <- 1.0 # FIIXMEE",
+#'   linters = todo_comment_linter(todo = "fiixmee")
 #' )
 #'
 #' lint(
-#'   text = "x <- TRUE # hack",
-#'   linters = todo_comment_linter(todo = c("todo", "fixme", "hack"))
+#'   text = "x <- TRUE # TOODOO(#1234): Fix this hack.",
+#'   linters = todo_comment_linter()
```
Comment (AshesITR) [inline R/todo_comment_linter.R:23]: Why this?

File: R/object_usage_linter.R (hunk 4 above)
Comment (AshesITR) [reply R/object_usage_linter.R:90]: Do we have a test for these?

File: R/object_usage_linter.R (hunk 5 above)
Comment (MichaelChirico) [reply R/object_usage_linter.R:212]: Let's do it more broadly as part of #2450.

**Comment by AshesITR** [comment]
Seems manageable to me. The comment fixes will probably be easy to review.

File: R/object_usage_linter.R (hunk 4 above)
Comment (MichaelChirico) [reply R/object_usage_linter.R:90]: Not sure... #2451

Comment (AshesITR) [reply R/object_usage_linter.R:90]: #2451

File: R/todo_comment_linter.R (hunk 6 above)
Comment (MichaelChirico) [reply R/todo_comment_linter.R:23]: NB: diff is screwing this up a bit. This is just deleting an example. I think that example is pretty redundant now that the earlier two examples also introduce custom `todo`. Earlier, the first two examples relied on defaults and so this example was showing something new; now, not so I think.

**Comment by AshesITR** [comment]
LGTM. Ping when you want to merge this (to avoid tons of "Update branch" merges with the current set of PRs

**Comment by MichaelChirico** [comment]
> LGTM. Ping when you want to merge this (to avoid tons of "Update branch" merges with the current set of PRs

g2g, btw, as long as there's no conflicts merge commits don't invalidate reviews. a bit hard to forecast which will cause conflicts OTOH.

**Review by AshesITR — APPROVED** [review]

**Pull request merged**